java/vector/src/main/java/org/apache/arrow/vector/DensityAwareVector.java

```diff
@@ -0,0 +1,57 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.vector;
+
+/**
+ * Vector that support density aware initial capacity settings.
+ * We use this for ListVector and VarCharVector as of now to
+ * control the memory allocated.
+ *
+ * For ListVector, we have been using a multiplier of 5
+ * to compute the initial capacity of the inner data vector.
+ * For deeply nested lists and lists with lots of NULL values,
+ * this is over-allocation upfront. So density helps to be
+ * conservative when computing the value capacity of the
+ * inner vector.
+ *
+ * For example, a density value of 10 implies each position in the
+ * list vector has a list of 10 values. So we will provision
+ * an initial capacity of (valuecount * 10) for the inner vector.
+ * A density value of 0.1 implies out of 10 positions in the list vector,
+ * 1 position has a list of size 1 and remaining positions are
```

Comment:
Sorry I don't understand this completely. Why is the 1 position that is not null has size == 1?

Reply:
Yeah I think it makes sense that density of 0.1 means 10% of the value is non-null. What I am not sure about is why the size of the non-null value has size 1 ? It seems `valuecount * density` is used for  both (1) number of non-null sub lists in the parent list (2) (average) length of the non-null sub lists. What if these two values are different?

Reply:
valueCount * density is used for computing the value capacity of the inner vector. IF the List vector has a valuecount of 10, we use the density to compute the target value count for the inner vector and < 1 value of density helps in provisioning keeping NULL values into account.

Reply:
Sounds like for a list of 10 values. These two have the same density == 1:
* 10 sub lists of size 1
* 1 sub list of size 10, 9 sub list of null

Is that correct understanding? The doc seems to fix the two cases so it's not very clear to me.

Reply:
The doc doesn't mix that. It clearly indicates the purpose of density:

For example, a
   *                density value of 10 implies each position in the list
   *                vector has a list of 10 values.
   *                A density value of 0.1 implies out of 10 positions in
   *                the list vector, 1 position has a list of size 1 and
   *                remaining positions are null (no lists) or empty lists.
   *                This helps in tightly controlling the memory we provision
   *                for inner data vector.

Reply:
What if I have a vector such that 1 out of 10 positions has a list of size 10 and remaining positions are null, what would the density be?

Reply:
Density is the average size of list per position in the List vector as mentioned in the doc. For your example, density is 1. I don't think it is a good idea to generalize the purpose or usage of density. The main purpose of density was to be conservative about the value capacity provisioned for inner vectors.

Reply:
I see. Thanks for the explanation.

> Density is the average size of list per position in the List vector

This is fine. 

>   density value of 10 implies each position in the list vector has a list of 10 values.

If I understand correctly, a density value of 10 can be either:
* 10 sub list of 10 values
* 1 sub list 100 values, 9 null sublists
* ...
As long as the average size of sub lists equals density.

Is that correct? If so, can we make it clear in the doc?

Reply:
I think you are trying to generalize the meaning of density whereas the list vector could be nested too. We propagate density down the tree. So here we just talk about the immediate inner vector.

Reply:
Which is why saying average list size implies the right meaning.

Reply:
Ok.